utils/src/main/kotlin/OrtProxySelector.kt

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2020 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.utils
+
+import java.io.IOException
+import java.net.InetSocketAddress
+import java.net.PasswordAuthentication
+import java.net.Proxy
+import java.net.ProxySelector
+import java.net.SocketAddress
+import java.net.URI
+
+typealias AuthenticatedProxy = Pair<Proxy, PasswordAuthentication?>
+
+const val DEFAULT_PROXY_PORT = 8080
+
+/**
+ * A proxy selector which supports dynamic addition and removal of proxies with optional password authentication.
+ */
+class OrtProxySelector(private val fallback: ProxySelector? = null) : ProxySelector() {
+    companion object {
+        private val NO_PROXY_LIST = listOf(Proxy.NO_PROXY)
+
+        /**
+         * Install this proxy selector as the global default. The previous selector is used as a fallback.
+         */
+        fun install(): OrtProxySelector {
+            val current = getDefault()
+            return if (current is OrtProxySelector) {
+                log.info { "Proxy selector is already installed." }
+                current
+            } else {
+                OrtProxySelector(current).also {
+                    setDefault(it)
+                    log.info { "Proxy selector was successfully installed." }
+                }
+            }
+        }
+
+        /**
+         * Uninstall this proxy selector, restoring the previous selector as the global default.
+         */
+        fun uninstall(): ProxySelector? {
+            val current = getDefault()
+            return if (current is OrtProxySelector) {
+                current.fallback.also {
+                    setDefault(it)
+                    log.info { "Proxy selector was successfully uninstalled." }
+                }
+            } else {
+                log.info { "Proxy selector is not installed." }
+                current
+            }
+        }
+    }
+
+    private val proxyAuthentication = mutableMapOf<Proxy, PasswordAuthentication?>()
+    private val proxyOrigins = mutableMapOf<String, MutableMap<String, MutableList<Proxy>>>()
```

Comment:
Is it required to allow arbitrary origin strings, or could this be an enum? If not, it should be documented somewhere, what an origin is.

Reply:
As `OrtProxySelector` does not know which origins could possibly be used in the future, I'd prefer to keep this as string. It's more or less a hint to humans / help for debugging anyway, but also to "manage" proxies on an origin / scope level. I've improved the docs for `addProxy()` in that regard.